### PR TITLE
[#IOCIT-378] Added error event tracking to ACS and Lollipop login middleware

### PR DESCRIPTION
**Pull request by michaeldisaro**
#### List of Changes

Added event tracking for errors during lollipop initialization on ACS and login middleware.

#### Motivation and Context

We need these events for some troubleshooting.

#### How Has This Been Tested?

- yarn build
- yarn lint
- yarn test

#### Types of changes

- [x] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/controllers/authenticationController.ts [hunk 1]
```diff
@@ -224,11 +234,20 @@ export default class AuthenticationController {
...
+            name: "lollipop.error.acs",
+            properties: {
+              assertion_ref: assertionRefToRevoke,
+              error: err,
+              fiscal_code: sha256(user.fiscal_code)
+            }
+          });
           log.error(
             "acs: error sending revoke message for previous assertionRef [%s]",
             err
```
Comment (gquadrati) [inline src/controllers/authenticationController.ts:253]: Prepend a message error that allow to discriminate what `lollipop.error.acs` event is this one.
We can reuse the log.error one.

File: src/controllers/authenticationController.ts [hunk 2]
```diff
@@ -297,9 +316,19 @@ export default class AuthenticationController {
...
           ),
-          TE.mapLeft(() =>
-            O.some(ResponseErrorInternal("Error Activation Lollipop Key"))
-          )
+          TE.mapLeft(error => {
+            this.appInsightsTelemetryClient?.trackEvent({
+              name: "lollipop.error.acs",
+              properties: {
+                assertion_ref: assertionRef,
+                error: error.message,
```
Comment (gquadrati) [inline src/controllers/authenticationController.ts:324]: ditto

File: src/controllers/authenticationController.ts [hunk 3]
```diff
@@ -4,6 +4,8 @@
  * the IDP.
  */
 
+/* eslint-disable sonarjs/no-duplicate-string */
```
Comment (gquadrati) [inline src/controllers/authenticationController.ts:7]: Can we remove this?

File: src/controllers/__tests__/authenticationController.test.ts [hunk 4]
```diff
@@ -695,6 +696,14 @@ describe("AuthenticationController#acs", () => {
       const response = await lollipopActivatedController.acs(validUserPayload);
       response.apply(res);
 
+      expect(mockTelemetryClient.trackEvent).toHaveBeenCalledWith({
+        name: "lollipop.error.acs",
+        properties: expect.objectContaining({
+          assertion_ref: anotherAssertionRef,
+          fiscal_code: sha256(aFiscalCode)
```
Comment (gquadrati) [inline src/controllers/__tests__/authenticationController.test.ts:703]: Can we check the `error` property too?

File: src/controllers/authenticationController.ts [hunk 5]
```diff
@@ -215,6 +218,13 @@ export default class AuthenticationController {
       : E.right(O.none);
 
     if (E.isLeft(errorOrMaybeAssertionRef)) {
+      this.appInsightsTelemetryClient?.trackEvent({
```
Comment (gquadrati) [inline src/controllers/authenticationController.ts:221]: Add test

File: src/controllers/authenticationController.ts [hunk 6]
```diff
@@ -224,11 +234,20 @@ export default class AuthenticationController {
...
       O.isSome(errorOrMaybeAssertionRef.right)
     ) {
+      const assertionRefToRevoke = errorOrMaybeAssertionRef.right.value;
       // Sending a revoke message for previous assertionRef related to the same fiscalCode
       // This operation is fire and forget
       this.lollipopParams.lollipopService
-        .revokePreviousAssertionRef(errorOrMaybeAssertionRef.right.value)
+        .revokePreviousAssertionRef(assertionRefToRevoke)
         .catch(err => {
+          this.appInsightsTelemetryClient?.trackEvent({
```
Comment (gquadrati) [inline src/controllers/authenticationController.ts:243]: Add test

**Review by gquadrati — COMMENTED** [review]
Just some small comments

File: src/controllers/authenticationController.ts (hunk 1 above)
Comment (michaeldisaro) [reply src/controllers/authenticationController.ts]: I added a message property, I think it's more clean. https://github.com/pagopa/io-backend/pull/996/commits/275e94a174fed76bf20736cb6040e924ad01a415

File: src/controllers/authenticationController.ts (hunk 5 above)
Comment (michaeldisaro) [reply src/controllers/authenticationController.ts:221]: done in https://github.com/pagopa/io-backend/pull/996/commits/275e94a174fed76bf20736cb6040e924ad01a415

File: src/controllers/authenticationController.ts (hunk 6 above)
Comment (michaeldisaro) [reply src/controllers/authenticationController.ts:243]: done in https://github.com/pagopa/io-backend/pull/996/commits/06eceba5db872dc0e45f3d08e6925e71a721e63e

File: src/controllers/__tests__/authenticationController.test.ts (hunk 4 above)
Comment (michaeldisaro) [reply src/controllers/__tests__/authenticationController.test.ts]: done https://github.com/pagopa/io-backend/pull/996/commits/06eceba5db872dc0e45f3d08e6925e71a721e63e

File: src/controllers/authenticationController.ts [hunk 7]
```diff
@@ -297,9 +327,19 @@ export default class AuthenticationController {
               )
             )
           ),
-          TE.mapLeft(() =>
-            O.some(ResponseErrorInternal("Error Activation Lollipop Key"))
-          )
+          TE.mapLeft(error => {
```
Comment (BurnedMarshal) [inline src/controllers/authenticationController.ts:330]: Split this event in two. The first inside the lollipop service (only the catch error and the response decode errors), the second instead inside the pipe of `setLollipopAssertionRefForUser`

Comment (michaeldisaro) [reply src/controllers/authenticationController.ts]: done https://github.com/pagopa/io-backend/pull/996/commits/acab1e4c85b2ffeb63e2296d387f274629178eab

**Review by gquadrati — APPROVED** [review]

**Review by gquadrati — APPROVED** [review]

**Pull request merged**